homeassistant/components/mqtt/mixins.py

```diff
@@ -868,7 +872,37 @@ async def async_removed_from_registry(self) -> None:
     @callback
     def add_to_platform_abort(self) -> None:
         """Abort adding an entity to a platform."""
+        hook_register: dict[tuple, Callable] = self.hass.data[
+            DATA_MQTT_DISCOVERY_REGISTRY_HOOKS
+        ]
+
+        async def _async_clear_discovery_topic_if_entity_removed(
+            discovery_data: dict[str, Any],
+            event: Event,
+        ) -> None:
+            """Clear the discovery topic when the entiy is removed."""
+            if event.data["action"] == "remove":
+                # publish empty payload to config topic to avoid re-adding
+                hass = async_get_hass()
+                discovery_hash: tuple[str, str] = discovery_data[ATTR_DISCOVERY_HASH]
+                discovery_prefix: str = hass.data[DATA_MQTT].conf[CONF_DISCOVERY_PREFIX]
+                topic = f"{discovery_prefix}/{discovery_hash[0]}/{discovery_hash[1].replace(' ','/')}/config"
+                hass.async_create_task(async_publish(hass, topic, "", retain=True))
+                hook_register.pop(discovery_hash)()
+
         if self._discovery_data:
+            discovery_hash: tuple = self._discovery_data[ATTR_DISCOVERY_HASH]
+            if discovery_hash not in hook_register:
```

Comment:
We want one listener per discovery_hash

Reply:
I don't think this is correct because we don't handle the case when an entity has a registry hook setup, and is then successfully added.

Maybe the cleanup could be done in `MqttDiscoveryUpdate.__init__` instead?

Reply:
If it is added successfully abort is not called. If it is added later, a reload will be performed and the hooks get cleaned up.